### ISSUE TITLE
no error when accessing `__name__` on object with no `__name__` attribute

### ISSUE THREAD

**Issue by DetachHead** [microsoft/pyright#7145]
```py
class Foo:
    ...

Foo().__name__ # no error, crashes at runtime
```
[playground](https://pyright-play.net/?strict=true&deprecateTypingAliases=true&enableExperimentalFeatures=true&reportCallInDefaultInitializer=true&reportImplicitOverride=true&reportImplicitStringConcatenation=true&reportPropertyTypeMismatch=true&reportMissingSuperCall=true&reportShadowedImports=true&reportUninitializedInstanceVariable=true&reportUnnecessaryTypeIgnoreComment=true&reportUnusedCallResult=true&code=MYGwhgzhAEBiD28BcBYAUNT0B0v3oXgAoBKbAfXIDswBbAU0ugGJor5p6AnL%2BLgGmjAukABb0YYAC7QuAVypSAlgyA)

**Comment by erictraut** [comment]
This is arguably a bug in the typeshed definition for `type`, which indicates that the `type` class has an instance variable named `__name__`. That means it's effectively a class variable for any class object created as an instance of `type`, which means it should be accessible to any instance of any class. So I think pyright is correctly interpreting the type information it is provided. It appears that the runtime is special-casing `__name__` so this attribute doesn't act like other metaclass instance variables. There's no way for the typeshed stub to indicate this special-case runtime behavior, so I don't see an obvious way to "fix this" in the stubs. 

Our choices are then to 1) ignore this false negative, or 2) add special-case logic in pyright to model the runtime behavior.

Up until now, I've chosen option 1. This is a false negative, after all, and it's a mistake that developers are unlikely to make in their code. Pyright's test cases even include this case with special comments indicating that the type checker behavior deviates from runtime behavior. However, in this case, I think there's a reasonable argument to be made in favor of adding the special casing (option 2).

I really dislike adding special-case behavior to pyright because behaviors appear inconsistent and it ties the type checker implementation to the typeshed stub implementation, which leads to fragility. In this particular case, the special-casing is pretty minimal, and I think it's less likely than other special-casing logic to run afoul of problems in the future. For that reason, I've gone ahead and made this change. It will be included in the next release.

**Comment by DetachHead** [comment]
should we instead try to make the fix in typeshed?

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.349, which I just published.
